cli/flox/src/commands/mod.rs

```diff
@@ -329,14 +343,210 @@ impl FloxArgs {
+            Commands::Additional(group) => group.handle(config, flox).await,
+            Commands::Internal(group) => group.handle(config, flox).await,
+        };
+
+        // This will print the update notification after output from a successful
+        // command but before an error is printed for an unsuccessful command.
+        // That's a bit weird,
+        // but I'm not sure it's worth a refactor.
+        match check_for_update_handle.await {
+            Ok(update_notification) => {
+                UpdateNotification::handle_update_result(update_notification);
+            },
+            Err(e) => debug!("Failed to check for CLI update: {}", display_chain(&e)),
+        }
+
+        result
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+struct LastUpdateNotification {
+    #[serde(with = "time::serde::iso8601")]
+    last_notification: OffsetDateTime,
+}
+
+#[derive(Debug, PartialEq)]
+struct UpdateNotification {
+    new_version: String,
+    notification_file: PathBuf,
+}
+
+#[derive(Debug, Error)]
+enum UpdateNotificationError {
+    /// If someone can't check for updates because of a network error, we'll
+    /// want to silently ignore it.
+    #[error("network error")]
+    Network(#[source] reqwest::Error),
+    /// If someone can't check for updates because of an IO error, we'll want to
+    /// silently ignore it.
+    #[error("IO error")]
+    Io(#[source] io::Error),
+    /// Other errors indicate something we didn't expect may have happened,
+    /// so we want to report it with Sentry.
+    #[error(transparent)]
+    WeMayHaveMessedUp(#[from] anyhow::Error),
+}
+
+impl UpdateNotification {
+    pub async fn check_for_and_print_update_notification(cache_dir: impl AsRef<Path>) {
+        Self::handle_update_result(Self::check_for_update(cache_dir).await)
+    }
+
+    /// If the user hasn't been notified of an update after
+    /// UPDATE_NOTIFICATION_EXPIRY time has passed, check for an update.
+    pub async fn check_for_update(
+        cache_dir: impl AsRef<Path>,
+    ) -> Result<Option<Self>, UpdateNotificationError> {
+        let notification_file = cache_dir.as_ref().join(UPDATE_NOTIFICATION_FILE_NAME);
+        Self::testable_check_for_update(
+            notification_file,
+            Self::get_latest_version,
+            UPDATE_NOTIFICATION_EXPIRY,
+        )
+        .await
+    }
+
+    /// If the user hasn't been notified of an update after `expiry` time has
+    /// passed, check for an update.
+    async fn testable_check_for_update<Fut>(
+        notification_file: PathBuf,
+        get_latest_version_function: impl Fn() -> Fut,
+        expiry: Duration,
+    ) -> Result<Option<Self>, UpdateNotificationError>
+    where
+        Fut: Future<Output = Result<String, UpdateNotificationError>>,
+    {
+        // Return early if we find a notification_file with a last_notification
+        // that hasn't expired
+        match tokio::fs::read_to_string(&notification_file).await {
```

Comment:
Suggestion: this read is likely to be performed in a single `read` syscall because the file is small, so there's likely no benefit to this particular call being `async`.

Reply:
Fixed